storage: prohibit splitting on Range.StartKey for non-SQL ranges

nvb opened the pull request:
Fixes #30396.

Before this change, we allowed splits on a range's StartKey as long as
that key had multiple MVCC versions. This change logically extends #24896
to fix this issue as well by preventing MVCCFindSplitKey from returning
the range's start key. In doing so, it also prevents MVCCFindSplitKey
from returning the first key in its range, even if it's not equal to the
range's StartKey. This is free to enforce and seems like the right thing
to do.

Release note: None

nvb committed:
storage: prohibit splitting on range.StartKey for non-SQL ranges

Fixes #30396.

Before this change, we allowed splits on a range's StartKey as long as
that key had multiple MVCC versions. This change logically extends #24896
to fix this issue as well by preventing MVCCFindSplitKey from returning
the range's start key. In doing so, it also prevents MVCCFindSplitKey
from returning the first key in its range, even if it's not equal to the
range's StartKey. This is free to enforce and seems like the right thing
to do.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/30445)
<!-- Reviewable:end -->

nvanbenschoten commented:
I confirmed that the backup from #30396 succeeds with this change.

benesch reviewed (approved):
Rare to see that the fix for something this obscure this actually makes the code cleaner. LGTM! Nice find.

Tiny nit: might be worth updating the big ol' comment in MVCCFindSplitKey to describe this edge case too. https://github.com/cockroachdb/cockroach/blob/0f3ee2b991d1c337630224a9511b5715ce7fc7c4/pkg/storage/engine/mvcc.go#L2687

benesch commented on pkg/storage/engine/mvcc.go:
```diff
@@ -2728,21 +2728,25 @@ func MVCCFindSplitKey(
 			return nil, err
 		}
-		minSplitKey = roachpb.RKey(firstRowKey.PrefixEnd())
+		minSplitKey = firstRowKey.PrefixEnd()
```
Wait, don't you want to drop the `.PrefixEnd` now that you compare against `minSplitKey` with `<` instead of `<=`?

benesch replied:
Oh, no, never mind. I confused myself. I do think that you're right that this is broken for interleaved tables.

nvb committed:
storage: allow splitting between a SQL row and its interleaved rows

Before this change, we prevented splits between the first SQL data row
and its interleaved rows. This commit fixes this issue.

Release note: None

nvanbenschoten commented:
Ok, this is ready for another review. I addressed the interleaved row issue, so I think it's all set now. @jordanlewis might also be interested in taking a look at this.

a-robinson reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 4 of 4 files at r1, 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30445#-:-LMscrfW1ejWhO0y1HGC:bg7igkv)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/storage/engine/engine.go, line 103 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/30445#-LMsZAW0EvkSLwTwEYgI:-LMsZAW0EvkSLwTwEYgJ:bdfe6xb) ([raw file](https://github.com/cockroachdb/cockroach/blob/485da095653ef3b57f934e46b54d9c46d7481501/pkg/storage/engine/engine.go#L103)):*
> ```Go
> 	// the split is roughly targetSize bytes. The returned key will never be
> 	// chosen from the key ranges listed in keys.NoSplitSpans and will always
> 	// sort equal to or after minSplitKey.
> ```

Should the two commits be squashed together? It seems odd to do something and then immediately undo it.
___
*[pkg/storage/engine/mvcc.go, line 2695 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/30445#-LMsZNja-pqqD08L2OfH:-LMsZNja-pqqD08L2OfI:bxadf39) ([raw file](https://github.com/cockroachdb/cockroach/blob/485da095653ef3b57f934e46b54d9c46d7481501/pkg/storage/engine/mvcc.go#L2695)):*
> ```Go
> 	// In addition, we must never return a split key that falls within a table
> 	// row. (Rows in tables with multiple column families are comprised of
> 	// multiple keys, one key per column family.) However, we do allows a split
> ```

s/allows/allow
___
*[pkg/storage/engine/mvcc.go, line 2696 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/30445#-LMsZY6v0tmw6rc8tk28:-LMsZY6v0tmw6rc8tk29:bcf5tzn) ([raw file](https://github.com/cockroachdb/cockroach/blob/485da095653ef3b57f934e46b54d9c46d7481501/pkg/storage/engine/mvcc.go#L2696)):*
> ```Go
> 	// row. (Rows in tables with multiple column families are comprised of
> 	// multiple keys, one key per column family.) However, we do allows a split
> 	// key that falls between a row and its interleaved rows.
> ```

Huh, TIL. It'd be nice if we had a preference against doing so other than in exceptional cases with massive rows in the non-interleaved table or a lot of data in the interleaved table, but that'd obviously be future work.


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> on the interleaved table bit. Nice tests, too!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30445#-:-LMsge6S5PuYmZC9XET7:b-wwph3g)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->

benesch reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30445)* status: :shipit: complete! 2 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
TFTRs!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30445#-:-LMyIkNPAPrpbwtCXcrB:b-tlx7ko)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale)

___
*[pkg/storage/engine/engine.go, line 103 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/30445#-LMsZAW0EvkSLwTwEYgI:-LMyIQhB0XWX9_jhRhdJ:b-gexvpe) ([raw file](https://github.com/cockroachdb/cockroach/blob/485da095653ef3b57f934e46b54d9c46d7481501/pkg/storage/engine/engine.go#L103)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Should the two commits be squashed together? It seems odd to do something and then immediately undo it.
</blockquote></details>

Yeah, there's an argument to be made that they should be squashed because there's a bit of churn in this interface, but they fix different bugs, so I think I'll keep them separate.

EDIT: I went through and removed the churn.
___
*[pkg/storage/engine/mvcc.go, line 2695 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/30445#-LMsZNja-pqqD08L2OfH:-LMyIbLk7DnqzIcmN6xN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/485da095653ef3b57f934e46b54d9c46d7481501/pkg/storage/engine/mvcc.go#L2695)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

s/allows/allow
</blockquote></details>

Done.
___
*[pkg/storage/engine/mvcc.go, line 2696 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/30445#-LMsZY6v0tmw6rc8tk28:-LMyIc5kCx1qajjZ-bL4:b2bt2f4) ([raw file](https://github.com/cockroachdb/cockroach/blob/485da095653ef3b57f934e46b54d9c46d7481501/pkg/storage/engine/mvcc.go#L2696)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Huh, TIL. It'd be nice if we had a preference against doing so other than in exceptional cases with massive rows in the non-interleaved table or a lot of data in the interleaved table, but that'd obviously be future work.
</blockquote></details>

Yeah, that's an interesting point. I wonder how useful a preference is without a guarantee though.


<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Fixed lint warning.

bors r+
